Add mainnet genesis configuration

AnnaShaleva opened the pull request:
Add genesis configuration for Neo X mainnet. @txhsl, @chenquanyu please review and ensure that contracts code is OK.

Can't be merged until all TODOs are resolved.

Close #64.

AnnaShaleva committed:
config: copy genesis_testnet.json to genesis_mainnet.json

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
config: add mainnet network ID

As discussed, it's set to be 47763 which in hex reassembles "bane".

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
config: update system contracts

Recompile system contracts based on the updated source code. Includes
update from #259, #260, #267.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented on config/genesis_mainnet.json:
```diff
@@ -33,9 +33,6 @@
   "alloc": {
-    "0xacC85FFb71f83b9bb264f6d64541926D375a6C1d": {
-      "balance": "1000000000000000000"
-    },
```
I don't think we need some side balances for mainnet, but maybe I'm wrong. @steven1227, could you confirm?

txhsl replied:
This is the bridge relayer address, if it has no balance to send the first transaction, then other GAS will be locked in bridge and treasury.

AnnaShaleva replied:
Good, reverted. Should we use the same address for mainnet?

txhsl replied:
It depends on the bridge configuration @gsmachado @mialbu.

mialbu replied:
@steven1227 Can you provide the address for the bridge relayer for mainnet?

AnnaShaleva commented on config/genesis_mainnet.json:
```diff
@@ -19,7 +19,7 @@
       "period": 10,
       "standbyValidators": [
-        "0xcbbeca26e89011e32ba25610520b20741b809007",
+        "0xcbbeca26e89011e32ba25610520b20741b809007", // TODO: update validators
```
@steven1227, @gsmachado please, provide StandBy validators addresses for mainnet.

AnnaShaleva commented on config/genesis_mainnet.json:
```diff
@@ -46,6 +46,7 @@
     },
     "0x1212000000000000000000000000000000000001": {
       "balance": "0",
+      // TODO: update *proxy* code? @txhsl, could you tell me how to update proxy code?
```
@txhsl, could you tell me where can I get the code of proxy contracts?

txhsl replied:
Please refer to https://github.com/OpenZeppelin/openzeppelin-contracts/blob/release-v5.0/contracts/proxy/ERC1967/ERC1967Proxy.sol.

This file was added by #7, but we didn't change it, so it was removed in #210.

AnnaShaleva commented on config/genesis_mainnet.json:
```diff
@@ -91,7 +92,7 @@
     },
     "0x1212000000000000000000000000000000000004": {
-      "balance": "1000000000000000000",
+      "balance": "1000000000000000000", // TODO: decide what is the initial amount of gas for Treasury contract
```
@steven1227, @gsmachado we need to decide what is the initial amount of GAS for Treasury contract.

mialbu replied:
🚨 This is not the treasury contract. This is the bridge contract. The treasury contract is `0x1212000000000000000000000000000000000006`.

AnnaShaleva replied:
Yes, TODO is placed in the wrong place. But we still need to decide on it.

mialbu replied:
> But we still need to decide on it.

Yes, of course. Just helping to make sure we don't modify the value at the wrong place here. 😅

chenquanyu replied:
> @steven1227, @gsmachado we need to decide what is the initial amount of GAS for Treasury contract.

100 million, same as T4.

AnnaShaleva commented on config/genesis_mainnet.json:
```diff
@@ -228,7 +229,7 @@
       }
     },
-    "0x1212100000000000000000000000000000000004": {
+    "0x1212100000000000000000000000000000000004": { // TODO: update Bridge and BridgeManagement implementation once they are ready
```
@mialbu what is approximate date when mainnet version of Bridge contracts is finished?

mialbu replied:
I'll create the release PR today. Since I want Shili and Krain to review it, maybe today or tomorrow we have the release and the byte code.

mialbu replied:
Waiting for https://github.com/bane-labs/bridge-evm-contracts/pull/175

mialbu commented on config/genesis_mainnet.json:
```diff
@@ -0,0 +1,263 @@
+      "code": "0x6080604052600a600c565b005b60186014601a565b6051565b565b6000604c7f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc546001600160a01b031690565b905090565b3660008037600080366000845af43d6000803e808015606f573d6000f35b3d6000fdfea26469706673582212205473baeda08227b037acf1d10bff816015a11f298917e2465da958f40bcefac664736f6c63430008180033",
+      "storage": {
+        "0x0000000000000000000000000000000000000000000000000000000000000064": "0x1212000000000000000000000000000000000005",
+        "0x000000000000000000000000000000000000000000000000000000000000006e": "0x016345785d8a0000",
+        "0x000000000000000000000000000000000000000000000000000000000000006f": "0x0de0b6b3a7640000",
+        "0x0000000000000000000000000000000000000000000000000000000000000070": "0x152d02c7e14af6800000",
+        "0x0000000000000000000000000000000000000000000000000000000000000071": "0x64",
```
@steven1227
For mainnet, are we keeping the following initial values for the Gas bridge? Otherwise, please tell me what values we should use.

Fee for withdrawing: **0.1 GAS**
Minimum withdrawal amount: **1 GAS**
Maximum withdrawal amount: **100'000 GAS**
Max Nr of Deposits per Relay-Tx: **100**

AnnaShaleva replied:
Is there any decision on this?

mialbu replied:
> Is there any decision on this?

@steven1227

steven1227 replied:
Confirm this

mialbu committed:
config: add bridge byte code to mainnet genesis file

txhsl commented on config/genesis_mainnet.json:
```diff
@@ -0,0 +1,263 @@
+        "0x0000000000000000000000000000000000000000000000000000000000000001": "0x07",
+        "0x0000000000000000000000000000000000000000000000000000000000000002": "0x0de0b6b3a7640000",
+        "0x0000000000000000000000000000000000000000000000000000000000000003": "0x027b46536c66c8e3000000",
+        "0x0000000000000000000000000000000000000000000000000000000000000004": "0x043c33c1937564800000",
```
~~We need `0x6c6b935b8bbd400000` if we include #275.~~

No, we will have new values for this.

AnnaShaleva replied:
Let's wait until #275 is merged and then update config.

AnnaShaleva replied:
@txhsl, the final decision from merged #275 was to reduce register fee from 20000 GAS to 2000 GAS and to use 50% as a service fee. So I suppose we still need to change this value from `0x043c33c1937564800000` to `0x6c6b935b8bbd400000`, right?

txhsl replied:
The latest news is we again reduce from 2000 GAS to 1000 GAS, but keep 50% as the service fee rate, and set the target vote amount to 6,000,000 GAS.

```json
"0x0000000000000000000000000000000000000000000000000000000000000000": "0x01",
"0x0000000000000000000000000000000000000000000000000000000000000001": "0x07", // 7
"0x0000000000000000000000000000000000000000000000000000000000000002": "0x0de0b6b3a7640000", // 1e18
"0x0000000000000000000000000000000000000000000000000000000000000003": "0x04f68ca6d8cd91c6000000", // 6e24
"0x0000000000000000000000000000000000000000000000000000000000000004": "0x3635c9adc5dea00000", // 1e21
"0x0000000000000000000000000000000000000000000000000000000000000005": "0xec40", // 60480
"0x0000000000000000000000000000000000000000000000000000000000000010": "0x07",
"0x0000000000000000000000000000000000000000000000000000000000000011": "0x07",
```

txhsl replied:
Included in #281.

txhsl committed:
config: add Multisig system contract byte code to mainnet genesis cfg

txhsl committed:
config: remove unused zero paddings

mialbu commented on config/genesis_mainnet.json:
```diff
@@ -0,0 +1,263 @@
+        "0x0000000000000000000000000000000000000000000000000000000000000069": "0x1212000000000000000000000000000000000006",
+        "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc": "0x1212100000000000000000000000000000000005",
+        "0x9016d09d72d40fdae2fd8ceac6b6234c7706214fd39c1cd1e609a0528c199300": "0xbb03c5030cac72e290ae185a8b9b375c58f7a9a6",
+        "0xf0c57e16840df040f15088dc2f81fe391c3923bec73e23a9662efc9c229c6a00": "0x01"
```
These values are not updated for mainnet yet afaik.

@steven1227 FYI

mialbu replied:
@steven1227 Can you provide the roles on the evm side here?

We need the addresses of the following roles:
- owner
- relayer
- validator 1-7
- securityGuard
- governor

mialbu replied:
**Owner** (CommitteeMultiSig)
Storage Key: `0x9016d09d72d40fdae2fd8ceac6b6234c7706214fd39c1cd1e609a0528c199300`
Value: `0x1212000000000000000000000000000000000007`

**Relayer** (NGD)
Storage Key: `0x0000000000000000000000000000000000000000000000000000000000000064`
Value: `0x0aa8095a4438d2006ee6aef706a012b5457c375f`

**Validator 1** (NGD)
Storage Key: `0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e94354`
Value: `0x96790618cdd975e18c29af43bc736324d2bfcc27`

**Validator 2** (NGD)
Storage Key: `0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e94355`
Value: `0x2c467ed2ec4e8ad1f5a74e884b517814445b563d`

**Validator 3** (lazynode)
Storage Key: `0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e94356`
Value: `0x38858eaa6b4424bbf855ec93857622ebdbb520d5`

**Validator 4** (NSPCC)
Storage Key: `0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e94357`
Value: `0x0dad0310606e25c5f384dcb49a2f709ff9e540fc`

**Validator 5** (AxLabs)
Storage Key: `0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e94358`
Value: `0x351926e55e6976c63a96c6eb84f8c969b475a499`

**Validator 6** (AxLabs)
Storage Key: `0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e94359`
Value: `0x522414d05c1c26b230f265b3b6317928df9b22b5`

**Validator 7** (AxLabs)
Storage Key: `0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e9435a`
Value: `0xaedf83da2983a851c38037a3b6a4e368ad0fd794`

**Governor** (AxLabs)
Storage Key: `0x0000000000000000000000000000000000000000000000000000000000000067`
Value: `0xb125b0b6566b2db33eb358e6293b6e08dfdeb87a`

**SecurityGuard** (NGD)
Storage Key: `0x0000000000000000000000000000000000000000000000000000000000000068`
SecurityGuard: `0xbb03c5030cac72e290ae185a8b9b375c58f7a9a6`

mialbu replied:
We'll need the addresses for the following bridge roles:

@steven1227 
- [x] owner -> what's the state of this in regards to https://github.com/bane-labs/bridge-evm-contracts/issues/176?
- [x] relayer
- [x] validator 1
- [x] validator 2
- [x] validator 3
- [x] security guard

@roman-khimov @AnnaShaleva 
- [x] validator 4

txhsl replied:
> owner -> what's the state of this in regards to https://github.com/bane-labs/bridge-evm-contracts/issues/176?

The contract was merged through https://github.com/bane-labs/go-ethereum/pull/279, and its genesis code is in https://github.com/bane-labs/go-ethereum/pull/281.

You can review this modification, and let's set the EVM bridge owner address to `0x1212000000000000000000000000000000000007`.

mialbu replied:
@AnnaShaleva All role assignments are updated above in my comment.

txhsl committed:
config: update MainNet register fee and vote target amount settings

Set Governance candidate registration fee to 1000 GAS and target vote
amount to 6000000 GAS. Also adjust the system contracts documentation
to include an updated MainNet Governance settings.

AnnaShaleva committed:
config: update StandBy validators list for MainNet genesis config

NGD:
0x34a3b2abb99b4c128acf61dcbbd1fcac0b161652
0x641ec1c538fa17e6ad8193c9b580f6850b114280
0xe3973f57e8a0aa312c1917ab0e6a05d8b6af6609
0xa61ac4a4f006f4fceeb72ee0012a2d3367168d10

LazyNode:
0xe6d1a9db6a0893926bd81c0ef93aaaa543c116f0

AxLabs:
0x4fe8af0dbb633283d8e9703668142fd130f2818d

NSPCC:
0x763452f65353fffe73d46539e51a6ddfc0e2c86a

The order is not important since this list is sorted by dBFT.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
config: fill in MainNet Bridge and BridgeManagement genesis allocations

1. Supply Bridge Relayer (0x0aa8095a4438d2006ee6aef706a012b5457c375f) with some
initial balance.

2. Keep Bridge fee settings the same as for T4 and in accordance with
https://github.com/bane-labs/go-ethereum/pull/271#discussion_r1682802213.

3. Add the following BridgeManagement allocations:

Owner (CommitteeMultiSig)
Storage Key: 0x9016d09d72d40fdae2fd8ceac6b6234c7706214fd39c1cd1e609a0528c199300
Value: 0x1212000000000000000000000000000000000007

Relayer (NGD)
Storage Key: 0x0000000000000000000000000000000000000000000000000000000000000064
Value: 0x0aa8095a4438d2006ee6aef706a012b5457c375f

Validator 1 (NGD)
Storage Key: 0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e94354
Value: 0x96790618cdd975e18c29af43bc736324d2bfcc27

Validator 2 (NGD)
Storage Key: 0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e94355
Value: 0x2c467ed2ec4e8ad1f5a74e884b517814445b563d

Validator 3 (lazynode)
Storage Key: 0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e94356
Value: 0x38858eaa6b4424bbf855ec93857622ebdbb520d5

Validator 4 (NSPCC)
Storage Key: 0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e94357
Value: 0x0dad0310606e25c5f384dcb49a2f709ff9e540fc

Validator 5 (AxLabs)
Storage Key: 0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e94358
Value: 0x351926e55e6976c63a96c6eb84f8c969b475a499

Validator 6 (AxLabs)
Storage Key: 0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e94359
Value: 0x522414d05c1c26b230f265b3b6317928df9b22b5

Validator 7 (AxLabs)
Storage Key: 0x46501879b8ca8525e8c2fd519e2fbfcfa2ebea26501294aa02cbfcfb12e9435a
Value: 0xaedf83da2983a851c38037a3b6a4e368ad0fd794

Governor (AxLabs)
Storage Key: 0x0000000000000000000000000000000000000000000000000000000000000067
Value: 0xb125b0b6566b2db33eb358e6293b6e08dfdeb87a

SecurityGuard (NGD)
Storage Key: 0x0000000000000000000000000000000000000000000000000000000000000068
SecurityGuard: 0xbb03c5030cac72e290ae185a8b9b375c58f7a9a6

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

mialbu commented:
The allocations for the bridge and bridge management contracts look good to me. 👍🏼

AnnaShaleva committed:
config: update Governance and Multisig bytecode for MainNet

Update in accordance with compiled bytecode received with compilation
config from ./contracts folder.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>
